dockstore-webservice/src/main/java/io/dockstore/webservice/resources/WorkflowResource.java

```diff
@@ -906,6 +950,355 @@ private void updateInfo(Workflow oldWorkflow, Workflow newWorkflow) {
+                    HttpStatus.SC_BAD_REQUEST);
+        }
+    }
+
+    /**
+     * Publish the deposit on Zenodo
+     * @param actionsApi Zenodo API for publishing deposits
+     * @param depositionID Zenodo's ID for the deposition
+     */
+    private Deposit publishDepositOnZenodo(ActionsApi actionsApi, int depositionID) {
+        Deposit publishedDeposit;
+        try {
+            publishedDeposit = actionsApi.publishDeposit(depositionID);
+        } catch (ApiException e) {
+            LOG.error("Could not publish DOI on Zenodo. Error is " + e.getMessage(), e);
+            throw new CustomWebApplicationException("Could not publish DOI on Zenodo."
+                    + " Error is " + e.getMessage(), HttpStatus.SC_BAD_REQUEST);
+        }
+        return publishedDeposit;
+    }
+
+    /**
+     * Register a Zenodo DOI for the workflow version
+     * @param workflow    workflow for which DOI is registered
+     * @param workflowVersion workflow version for which DOI is registered
+     */
+    private void registerZenodoDOIForWorkflow(String zenodoAccessToken, Workflow workflow, WorkflowVersion workflowVersion) {
+        ApiClient zendoClient = new ApiClient();
+        // for testing, either 'https://sandbox.zenodo.org/api' or 'https://zenodo.org/api' is the first parameter
+        String zenodoUrlApi = zenodoUrl + "/api";
+        zendoClient.setBasePath(zenodoUrlApi);
+        zendoClient.setApiKey(zenodoAccessToken);
+
+        DepositsApi depositApi = new DepositsApi(zendoClient);
+        ActionsApi actionsApi = new ActionsApi(zendoClient);
+
+        Deposit deposit = new Deposit();
+        Deposit returnDeposit;
+
+        checkForExistingDOIForWorkflowVersion(workflowVersion);
+
+        String latestWorkflowVersionDOIURL = getAnExistingDOIForWorkflow(workflow);
+
+        int depositionID = 0;
+        DepositMetadata depositMetadata = null;
+
+        if (latestWorkflowVersionDOIURL == null || latestWorkflowVersionDOIURL.isEmpty()) {
+            try {
+                // No DOI has been assigned to any version of the workflow yet
+                // So create a new deposit which will enable creation of a new
+                // concept DOI and new version DOI
+                returnDeposit = depositApi.createDeposit(deposit);
+                depositionID = returnDeposit.getId();
+                depositMetadata = returnDeposit.getMetadata();
+
+                fillInMetadata(depositMetadata, workflow, workflowVersion);
+
+            } catch (ApiException e) {
+                LOG.error("Could not create deposition on Zenodo. Error is " + e.getMessage(), e);
+                throw new CustomWebApplicationException("Could not create deposition on Zenodo. "
+                        + "Error is " + e.getMessage(), HttpStatus.SC_BAD_REQUEST);
+            }
+        } else {
+            String depositIdStr = latestWorkflowVersionDOIURL.substring(latestWorkflowVersionDOIURL.lastIndexOf(".") + 1).trim();
```

Comment:
Could you please show a sample DOI in a comment? It's hard to visualize what this and line 1275 are doing.

Does a DOI always have a `.` and a `/`? Probably, just making sure because this line and line 1275 assume so.

Reply:
done